### PR TITLE
add envvar default handler

### PR DESCRIPTION
**Pull request by rmorshea**
substitute for #158 
closes #157 

see [test](https://github.com/rmorshea/traitlets/blob/dcd9e1db4717e1fac978d3fba999b174afd3fb69/traitlets/tests/test_traitlets.py#L179) / [doc](https://github.com/rmorshea/traitlets/blob/dcd9e1db4717e1fac978d3fba999b174afd3fb69/traitlets/traitlets.py#L868) for usage

### CODE REVIEW

File: traitlets/tests/test_traitlets.py [hunk 1]
```diff
@@ -176,6 +176,28 @@ def _default_x(self):
...
+    def test_envvar_dynamic_initializer(self):
+        import os
+
+        os.environ['MY_ENVVAR'] = 'true'
+
+        class A(HasTraits):
+            b = Bool()
+
+            @envvar_default('b', 'MY_ENVVAR')
+            def _b_from_envvar(self, init):
```
Comment (minrk) [inline traitlets/tests/test_traitlets.py]: What is `init` here? Shouldn't it be the value of the environment variable?

Comment (rmorshea) [inline traitlets/tests/test_traitlets.py]: I was working from the idea that the handlers should generally accept a 'bunch' object, or nothing, so I chose to include the trait in the 'init' bunch. I couldn't really think of a better name than 'init' so if there's something better I'm open to it.

**Comment by minrk** [comment]
While the decorator gives you the most control, needing to define a method means this doesn't really improve the situation over the status quo, where today you might have:

```python
class C(Configurable):
    a = Unicode().tag(config=True)
    @default('a')
    def _a_default(self):
        return os.environ.get('A_ENV', '')
```

whereas with this, it becomes:

```python
class C(Configurable):
    a = Unicode().tag(config=True)
    @envvar_default('a', 'A_ENV')
    def _a_default(self, init):
        return init.value or ''
```

which requires a new import, understanding a new construct, etc.

I'm not sure this is worth adding, unless it can be done via *just* a tag, at least sometimes, e.g.:

```python
class C(Configurable):
    a = Unicode().tag(config=True, envvar='A_ENV')
```

Then, *optionally*, a validate method could be used to handle any validation/casting/etc.

**Comment by Carreau** [comment]
Thanks @rmorshea !

I think one of the biggest point that could help get that in would be help autogeneration. 
It's relatively hard to keep docs in Sync, or even just write the docs when you use envvars.
Here I guess we could pull that out automatically !

I'm also wondering if it **has to** be on core. It looks here like we could make it a separate package, which would allow to : 
  - iterate on it
  - Bring functionality older traitlets versions. 

Though before diving into that, and want to make sure we want to move forward with this;
I tend to agree with @minrk, and the usecase seem to be really narrow, and I'm not convinced this will be that much used. Though I might not be the best to judge. 

I think we should need to be extra careful for the precedence rule of ENV var vs config, vs command line argument as well, and make sure we agree with the order, as well as the signification of having an env variable unset vs having an empty env variable. I can see also issue with casting as most env var are strings. 

I think @fperez  is often of help in these case. Can you look at this as well as related issue ? We tend to come back and forth on that and pull in many direction.

**Comment by rmorshea** [comment]
@minrk I think the fact that envvar defaults are only applicable to traits of `Configurables` makes `trait.tag(envvar='x')` very awkward to implement as a part of `TraitType` since it's available to all `HasTraits` subclasses. I'd argue the feature isn't not worth implementing if we need to do it with `tag`.

**Comment by rmorshea** [comment]
@Carreau I'll have to familiarize myself more with `sphinx` before I can cover something like #210, but I can make that my next project with `traitlets` since I've already put work into autodocs with #131.

As for where the pr belongs in traitlets? I'm not really too sure. Perhaps @fperez can help with that, or the original group involved with #157.

**Comment by Carreau** [comment]
> @Carreau I'll have to familiarize myself more with sphinx

I was more thinking of `--help-all` which woudl likely require to only patch a [single method in `configurable.py`](https://github.com/ipython/traitlets/blob/9988734c6e65f73e87aeabdcd689711c2aead188/traitlets/config/configurable.py#L205-L218)  to be aware of the env varaible, and basically just happend a templated paragraph. Sphinx is quite a complex beast.

**Comment by rmorshea** [comment]
@Carreau, given that. I think #131, makes [strong progress](https://github.com/rmorshea/traitlets/blob/doc_gen/traitlets/utils/docgen.py#L100-L124) towards that goal.

**Pull request closed without merging**